sdk/pinocchio/src/sysvars/slot_hashes/mod.rs

```diff
@@ -0,0 +1,443 @@
+
+/// Validates SlotHashes data format and returns the entry count.
+#[inline]
+fn parse_and_validate_data(data: &[u8]) -> Result<usize, ProgramError> {
+    // Need at least the 8-byte length prefix.
+    if data.len() < NUM_ENTRIES_SIZE {
+        return Err(ProgramError::AccountDataTooSmall);
+    }
+
+    let num_entries = unsafe { read_entry_count_from_bytes_unchecked(data) };
+
+    validate_slothashes_constraints(data.len(), Some(num_entries))
+}
+
+impl SlotHashEntry {
+    /// Returns the slot number as a u64.
+    #[inline(always)]
+    pub fn slot(&self) -> Slot {
+        u64::from_le_bytes(self.slot_le)
+    }
+}
+
+impl<T: Deref<Target = [u8]>> SlotHashes<T> {
+    /// Validates that a buffer is properly sized for SlotHashes data.
+    ///
+    /// Checks that the buffer length is 8 + (N * 40) for some N ≤ 512.
+    #[inline(always)]
+    pub(crate) fn validate_buffer_size(buffer_len: usize) -> Result<(), ProgramError> {
+        validate_slothashes_constraints(buffer_len, None)?;
+        Ok(())
+    }
+
+    /// Creates a `SlotHashes` instance from arbitrary data with full validation.
+    ///
+    /// This constructor performs comprehensive validation of the data format
+    /// including length prefix, entry count bounds, and buffer size requirements.
+    /// Does not validate that entries are sorted in descending order.
+    #[inline(always)]
+    pub fn new(data: T) -> Result<Self, ProgramError> {
+        let len = parse_and_validate_data(&data)?;
+        Ok(unsafe { Self::new_unchecked(data, len) })
+    }
+
+    /// Creates a `SlotHashes` instance directly from a data container and entry count.
+    /// Important: provide a valid len. Whether or not len is assumed to be
+    /// the constant 20_488 (512 entries) is up to caller.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not check the validity of the data or count.
+    /// The caller must ensure:
+    /// 1. The underlying byte slice in `data` represents valid SlotHashes data
+    ///    (length prefix + entries, where entries are sorted in descending order by slot).
+    /// 2. `len` is the correct number of entries (≤ MAX_ENTRIES), matching the prefix.
+    /// 3. The data slice contains at least `NUM_ENTRIES_SIZE + len * ENTRY_SIZE` bytes.
+    ///
+    #[inline(always)]
+    pub unsafe fn new_unchecked(data: T, len: usize) -> Self {
+        debug_assert!(len <= MAX_ENTRIES && data.len() >= NUM_ENTRIES_SIZE + len * ENTRY_SIZE);
+
+        SlotHashes { data, len }
+    }
+
+    /// Gets the number of entries stored in this SlotHashes instance.
+    /// Performs validation checks and returns the entry count if valid.
+    #[inline(always)]
+    pub fn get_entry_count(&self) -> Result<usize, ProgramError> {
+        let data_entry_count = read_entry_count_from_bytes(&self.data).unwrap_or(0);
+        if data_entry_count != self.len {
+            return Err(ProgramError::InvalidArgument);
+        }
+        Ok(self.len)
+    }
```

Comment:
`SlotHashes` does not specify what invariants should hold for the `data` field.
But `new()` calls `parse_and_validate_data(&data)` and assigns the computed lengths into the `len` field.
Which seems to suggest that the `len` field should always be valid.
If the above is true, this getter can be just returning `self.len` without a failure case.

It would also make `get_entry_count_unchecked()` redundant.

Reply:
As it stands right now, `new_unchecked` does not make that promise (see earlier https://github.com/anza-xyz/pinocchio/pull/152#discussion_r2116720695). AFAIK while it's invariant on mainnet that this will be 512, that might not be true in other contexts?

Reply:
Yes, this makes sense to me:

new_unchecked() will assume golden mainnet length.

Other contexts which might have <512 entries can either use new(), or zerofill up to 512 and use new_unchecked().

Reply:
Addressed in af5cf6e931e797547eff92fc748bd51c0090d96d, thanks. Removing support for shorter lengths also saved some CUs on some paths which handled non-standard lengths. Thanks!